File: sdk/php/SmartField.php
<?php
/**
 * SmartField Server SDK for PHP
 *
 * Usage:
 *   require_once 'SmartField.php';
 *   $sf = new SmartField();
 *   $sf->init();
 *
 *   // Laravel route
 *   Route::get('/sf-key', fn() => response()->json($sf->getPublicKey()));
 *   Route::post('/login', function(Request $r) use ($sf) {
 *       $email = $sf->decrypt($r->input('email'));
 *       $password = $sf->decrypt($r->input('password'));
 *   });
 */

class SmartField {
    private $privateKey = null;
    private $publicKey = null;
    private $publicKeyJwk = null;
    private $keysDir = null;

    public function init($keysDir = '.smartfield') {
        $this->keysDir = $keysDir;
        $privPath = $keysDir . '/private.pem';
        $pubPath = $keysDir . '/public.json';

        if (file_exists($privPath) && file_exists($pubPath)) {
            $this->privateKey = openssl_pkey_get_private(file_get_contents($privPath));
            $this->publicKeyJwk = json_decode(file_get_contents($pubPath), true);
            echo "[SmartField] Keys loaded from $keysDir\n";
            return;
        }

        // Generate new keys
        $config = [
            'private_key_bits' => 2048,
            'private_key_type' => OPENSSL_KEYTYPE_RSA,
        ];
        $res = openssl_pkey_new($config);
        openssl_pkey_export($res, $privPem);
        $details = openssl_pkey_get_details($res);

        $this->privateKey = $res;

        // Save keys
        if (!is_dir($keysDir)) mkdir($keysDir, 0700, true);
        file_put_contents($privPath, $privPem);
        chmod($privPath, 0600);

        // Public key as JWK
        $this->publicKeyJwk = [
            'kty' => 'RSA',
            'alg' => 'RSA-OAEP-256',
            'ext' => true,
            'key_ops' => ['encrypt'],
            'n' => $this->base64UrlEncode($details['rsa']['n']),
            'e' => $this->base64UrlEncode($details['rsa']['e']),
        ];
        file_put_contents($pubPath, json_encode($this->publicKeyJwk, JSON_PRETTY_PRINT));

        // Add to .gitignore
        $gitignore = getcwd() . '/.gitignore';
        if (file_exists($gitignore) && strpos(file_get_contents($gitignore), '.smartfield') === false) {
            file_put_contents($gitignore, "\n# SmartField keys\n.smartfield/\n", FILE_APPEND);
        }

        echo "[SmartField] New keys generated in $keysDir\n";
    }

    public function getPublicKey() {
        if (!$this->publicKeyJwk) throw new Exception('[SmartField] Not initialized');
        return $this->publicKeyJwk;
    }

    public function decrypt($encryptedPayload) {
        if (!$this->privateKey) throw new Exception('[SmartField] Not initialized');
        if (empty($encryptedPayload)) return '';

        $payload = json_decode(base64_decode($encryptedPayload), true);

        $iv = base64_decode($payload['iv']);
        $encryptedKey = base64_decode($payload['key']);
        $encryptedData = base64_decode($payload['data']);

        // Decrypt AES key with RSA-OAEP
        $aesKeyRaw = '';
        if (!openssl_private_decrypt($encryptedKey, $aesKeyRaw, $this->privateKey, OPENSSL_PKCS1_OAEP_PADDING)) {
            throw new Exception('[SmartField] RSA decryption failed');
        }

        // Decrypt data with AES-256-GCM
        // GCM tag is last 16 bytes of encrypted data
        $tagLength = 16;
        $tag = substr($encryptedData, -$tagLength);
        $ciphertext = substr($encryptedData, 0, -$tagLength);

        $decrypted = openssl_decrypt($ciphertext, 'aes-256-gcm', $aesKeyRaw, OPENSSL_RAW_DATA, $iv, $tag);

        if ($decrypted === false) {
            throw new Exception('[SmartField] AES decryption failed');
        }

        return $decrypted;
    }

    public function decryptFields($data) {
        if (!$this->privateKey) throw new Exception('[SmartField] Not initialized');

        $result = [];
        foreach ($data as $key => $value) {
            if (is_string($value) && strlen($value) > 50) {
                try {
                    $result[$key] = $this->decrypt($value);
                } catch (Exception $e) {
                    $result[$key] = $value;
                }
            } else {
                $result[$key] = $value;
            }
        }
        return $result;
    }

    private function base64UrlEncode($data) {
        return rtrim(strtr(base64_encode($data), '+/', '-_'), '=');
    }
}

File: sdk/go/smartfield.go
// Package smartfield provides server-side decryption for SmartField components.
// ============================================================================
// SmartField Server SDK for Go — v2.7.0
// ============================================================================
//
// Copyright (c) 2026 SmartField — MIT License
// Website:  https://smartfield.dev
// Docs:     https://smartfield.dev/docs
// Support:  support@smartfield.dev
//
// Decrypts data encrypted by the <smart-field> browser component.
// Uses Go standard crypto library (crypto/rsa + crypto/aes). Zero dependencies.
// Keys are generated and stored locally. SmartField never sees your data.
//
// API:
//   sf := smartfield.New()
//   sf.Init()                    — Generate or load RSA-2048 keys
//   sf.GetPublicKey()            — Return public key as JWK map
//   sf.Decrypt(payload)          — Decrypt a single encrypted value
//   sf.DecryptFields(fields)     — Decrypt all encrypted fields
//
// Encryption: AES-256-GCM (NIST SP 800-38D) + RSA-OAEP-2048 (NIST SP 800-56B)
// ============================================================================
package smartfield
import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"crypto/rsa"
	"crypto/sha256"
	"crypto/x509"
	"encoding/base64"
	"encoding/json"
	"encoding/pem"
	"fmt"
	"math/big"
	"os"
	"path/filepath"
)
type SmartField struct {
	privateKey *rsa.PrivateKey
	publicKey  *rsa.PublicKey
	keysDir    string
}
type payload struct {
	V    int    `json:"v"`
	IV   string `json:"iv"`
	Key  string `json:"key"`
	Data string `json:"data"`
}
func New() *SmartField {
	return &SmartField{}
}
func (sf *SmartField) Init(keysDir ...string) error {
	dir := ".smartfield"
	if len(keysDir) > 0 {
		dir = keysDir[0]
	}
	sf.keysDir = dir
	privPath := filepath.Join(dir, "private.pem")
	pubPath := filepath.Join(dir, "public.pem")
	// Load existing keys
	if _, err := os.Stat(privPath); err == nil {
		privPem, err := os.ReadFile(privPath)
		if err != nil {
			return err
		}
		block, _ := pem.Decode(privPem)
		key, err := x509.ParsePKCS8PrivateKey(block.Bytes)
		if err != nil {
			return err
		}
		sf.privateKey = key.(*rsa.PrivateKey)
		sf.publicKey = &sf.privateKey.PublicKey
		fmt.Printf("[SmartField] Keys loaded from %s\n", dir)
		return nil
	}
	// Generate new keys
	privateKey, err := rsa.GenerateKey(rand.Reader, 2048)
	if err != nil {
		return err
	}
	sf.privateKey = privateKey
	sf.publicKey = &privateKey.PublicKey
	os.MkdirAll(dir, 0700)
	// Save private key
	privBytes, err := x509.MarshalPKCS8PrivateKey(privateKey)
	if err != nil {
		return err
	}
	privPem := pem.EncodeToMemory(&pem.Block{Type: "PRIVATE KEY", Bytes: privBytes})
	os.WriteFile(privPath, privPem, 0600)
	// Save public key
	pubBytes, err := x509.MarshalPKIXPublicKey(&privateKey.PublicKey)
	if err != nil {
		return err
	}
	pubPem := pem.EncodeToMemory(&pem.Block{Type: "PUBLIC KEY", Bytes: pubBytes})
	os.WriteFile(pubPath, pubPem, 0644)
	fmt.Printf("[SmartField] New keys generated in %s\n", dir)
	return nil
}
func (sf *SmartField) GetPublicKey() map[string]interface{} {
	return map[string]interface{}{
		"kty":     "RSA",
		"alg":     "RSA-OAEP-256",
		"ext":     true,
		"key_ops": []string{"encrypt"},
		"n":       base64URLEncode(sf.publicKey.N.Bytes()),
		"e":       base64URLEncode(big.NewInt(int64(sf.publicKey.E)).Bytes()),
	}
}
func (sf *SmartField) Decrypt(encryptedPayload string) (string, error) {
	if encryptedPayload == "" {
		return "", nil
	}
	jsonBytes, err := base64.StdEncoding.DecodeString(encryptedPayload)
	if err != nil {
		return "", fmt.Errorf("base64 decode failed: %w", err)
	}
	var p payload
	if err := json.Unmarshal(jsonBytes, &p); err != nil {
		return "", fmt.Errorf("JSON parse failed: %w", err)
	}
	iv, _ := base64.StdEncoding.DecodeString(p.IV)
	encKey, _ := base64.StdEncoding.DecodeString(p.Key)
	encData, _ := base64.StdEncoding.DecodeString(p.Data)
	// Decrypt AES key with RSA-OAEP
	aesKeyRaw, err := rsa.DecryptOAEP(sha256.New(), rand.Reader, sf.privateKey, encKey, nil)
	if err != nil {
		return "", fmt.Errorf("RSA decrypt failed: %w", err)
	}
	// Decrypt data with AES-GCM
	block, err := aes.NewCipher(aesKeyRaw)
	if err != nil {
		return "", fmt.Errorf("AES cipher failed: %w", err)
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", fmt.Errorf("GCM failed: %w", err)
	}
	decrypted, err := gcm.Open(nil, iv, encData, nil)
	if err != nil {
		return "", fmt.Errorf("AES decrypt failed: %w", err)
	}
	return string(decrypted), nil
}
func (sf *SmartField) DecryptFields(fields map[string]string) map[string]string {
	result := make(map[string]string)
	for key, value := range fields {
		if len(value) > 50 {
			dec, err := sf.Decrypt(value)
			if err == nil {
				result[key] = dec
			} else {
				result[key] = value
			}
		} else {
			result[key] = value
		}
	}
	return result
}
func base64URLEncode(data []byte) string {
	return base64.RawURLEncoding.EncodeToString(data)
}
